yield_plugin/src/yield_plugin.cpp

```diff
@@ -659,22 +662,35 @@ namespace yield_plugin
         double vehicle_downtrack = wm_->routeTrackPos(vehicle_point).downtrack;
         double object_downtrack = wm_->routeTrackPos(object_point).downtrack;
 
-        if (vehicle_downtrack > object_downtrack + config_.vehicle_length / 2)  // if half a length of the vehicle past, it is considered behind
+        if (vehicle_downtrack > object_downtrack + config_.vehicle_length / 2) // if half a length of the vehicle past, it maybe considered behind
```

Comment:
The vehicle's down track is based on the position of its rear axle, right? If so, why are we moving the value forward by half a vehicle length? This seems like it would cause issues when an obstacle is near the vehicle's rear.

Reply:
2. Yes, it is based on the rear-axle. We can manually check it. Consider such scenarios (veh_length is 4 meters): 
- vehicle_downtrack = 80m and object_downtrack 82m.  It means ped is 2 meters in front of the back-axle. `if statement` goes: 80 > 82 + 2 is false, so vehicle is NOT passed. Correct
- vehicle_downtrack = 82m and object_downtrack 82m.  It means ped is at the back-axle. `if statement` goes 82 > 82 + 2 is false, so vehicle is NOT passed. Correct
- vehicle_downtrack = 84m and object_downtrack 82m.  It means ped is at the 1/2 veh length passed the rear-axle (1/2 veh len is more than distance between rear-axle and rear bumper). `if statement` goes 84 > 82 + 2 is false, so vehicle is NOT passed. Incorrect-> I guess I could change the expression to `>=`

Reply:
> * vehicle_downtrack = 84m and object_downtrack 82m.  It means ped is at the 1/2 veh length passed the rear-axle (1/2 veh len is more than distance between rear-axle and rear bumper for all cars except trucks maybe). `if statement` goes 84 > 82 + 2 is `false`, so vehicle is NOT passed. Incorrect-> I guess I could change the expression to `>=` . Since the `if statement` here should say `true`

It's unclear (to me at least) from the code alone that we modify the down track to accommodate for the vehicle overhang. I suggest changing

```cpp
if (vehicle_downtrack > object_downtrack + config_.vehicle_length / 2)
```

to 

```cpp
if (object_downtrack < vehicle_downtrack - config_.vehicle_length / 2)
```

and adding a clarifying comment about how we need to compensate for the vehicle's body overhang and that half the vehicle's total length is a conservative value.

Reply:
> I am hesitant to move the obstacle_behind_filter out of the function. It is mainly to avoid another 2 for loops to get data we already have.

We can avoid doing another two loops by refactoring the `get_collision_time()` function. Consider the following alternative:

```cpp
struct get_collision_result {
  rclcpp::Time collision_time;  // possibly unneeded
  lanelet::BasicPoint2d point1;
  lanelet::BasicPoint2d point2;
};

std::optional<get_collision_result> get_collision(/* traj1 */, /* traj2 */);

if (const auto result{get_collision(obstacle_trajectory, vehicle_trajectory)}) {
  const auto obstacle_downtrack{wm_->routeTrackPos(result.point1).downtrack};
  const auto vehicle_downtrack{wm->routeTrack(result.point2).downtrack};

  if (obstacle_behind_vehicle(obstacle_id, obstacle_downtrack, vehicle_downtrack) {
    // ...
  }
}
```

This is a similar architecture to the [`std::ranges::mismatch()`](https://en.cppreference.com/w/cpp/algorithm/ranges/mismatch) function.

> And having obstacle_behind_filter inside the get_collision_time makes sense because without it, get_collision_time returns wrong or incomplete result.

The `get_collision_time()` function's only job is to determine if there is a collision between two trajectories. Noisy trajectories mean we will get some false positives, but `get_collision_time()` doesn't need to know or care about that context. That's the caller's responsibility. From the function's perspective, the two provided trajectories resulted in a collision.

> The user would have to know to call this extra edge case handler function to get correct result.

The users in this case are downstream or service-calling ROS nodes. The code we are discussing is an implementation detail, so it is our responsibility to properly handle the edge case.

Reply:
Okay I incorporated your suggestions and added unit test for the new function.